### PR TITLE
Set `DotNetReleaseShipping` property in `manifest.json`

### PR DESCRIPTION
**Pull request by MilenaHristova**
Propagating `DotNetReleaseShipping` from `MergedManifest.xml` to assets in `manifest.json`

Issue: https://github.com/dotnet/release/issues/740

### Release Note Category
- [x] Feature changes/additions 
- [ ] Bug fixes
- [ ] Internal Infrastructure Improvements

### Release Note Description
Set `DotNetReleaseShipping` property in `manifest.json` in gather-drop

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs [hunk 1]
```diff
@@ -65,7 +71,8 @@ public class ManifestHelper
                         new
                         {
                             name = asset.Asset.Name,
-                            origin = assetOriginMap.TryGetValue(asset.Asset.Name, out var origin) ? origin : null,
+                            origin = assetReleaseMetadataMap.TryGetValue(asset.Asset.Name, out var data) ? data.Origin : null,
+                            dotnetReleaseShipping = data != null && data.DotNetReleaseShipping,
```
Comment (dkurepa) [inline src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs:75]:
```suggestion
                            dotnetReleaseShipping = data?.DotNetReleaseShipping,
```
does this work? not sure

File: src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs [hunk 2]
```diff
@@ -164,21 +173,34 @@ private static List<DownloadedAsset> SelectMergedManifestAssets(IEnumerable<Down
...
+
+    private class AssetReleaseMetadata
+    {
+        public AssetReleaseMetadata(string origin, bool dotNetReleaseShipping)
+        {
+            Origin = origin;
+            DotNetReleaseShipping = dotNetReleaseShipping;
+        }
+
+        public string Origin { get; set; }
```
Comment (dkurepa) [inline src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs:202]: maybe we can make both of these `{ get; init; }` so they're immutable. Doesn't look like we're changing them after initialization

File: src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs [hunk 3]
```diff
@@ -164,21 +173,34 @@ private static List<DownloadedAsset> SelectMergedManifestAssets(IEnumerable<Down
...
+                    assetMetadataMap.Add(assetName, new AssetReleaseMetadata(assetOrigin, dotNetReleaseShipping));
                 }
             }
         }
 
-        return assetOriginMap;
+        return assetMetadataMap;
+    }
+
+    private class AssetReleaseMetadata
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs:194]: This whole class could be 1 line btw:

```csharp
private record AssetReleaseMetadata(string Origin, bool DotNetReleaseShipping);
```

File: src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs [hunk 4]
```diff
@@ -156,6 +164,7 @@ private static List<DownloadedAsset> SelectMergedManifestAssets(IEnumerable<Down
                     continue;
 
                 string? assetOrigin = asset.Attribute("Origin")?.Value;
+                bool dotNetReleaseShipping = asset.Attribute("DotNetReleaseShipping")?.Value == "true";
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs:167]: Maybe compare and ignore case?

**Review by premun — APPROVED** [review]
LGTM just consider using the `record` down there

File: src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs (hunk 1 above)
Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs:75]: I don't know, for me it's really hard to understand this syntax

Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs:75]: Then keep it, it's not a big deal at all

File: src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs (hunk 3 above)
Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs]: updated

File: src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs (hunk 2 above)
Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs]: I'll  use a `record`

**Review by premun — APPROVED** [review]

File: src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs [hunk 5]
```diff
@@ -164,21 +173,23 @@ private static List<DownloadedAsset> SelectMergedManifestAssets(IEnumerable<Down
                 // it is taken from the Build.Name, which corresponds to the repository name.
                 assetOrigin ??= repoName;
 
-                if (assetOriginMap.TryGetValue(assetName, out var existingAssetOrigin))
+                if (assetMetadataMap.TryGetValue(assetName, out var existingAssetMetadata))
                 {
-                    if (existingAssetOrigin != assetOrigin)
```
Comment (MilenaHristova) [inline src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs:169]: I'm not sure about this condition - in which case can we have the same asset coming from different repos @andriipatsula ? If I understand correctly now the 2nd occurence of the asset will have a `null` origin, but what if the second one is shipping. Then maybe we should pick the shipping one

Comment (andriipatsula) [reply src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs:169]: I was/am not sure if we will have just a single MergedManifest.xml file or multiple MergedManifests for each repo, plus a single one that combines all repo-level MergedManifests. As I am reading all "MergedManifests.xml" file in the drop `List<DownloadedAsset> mergedManifestAssets = SelectMergedManifestAssets(downloadedBuilds);`

We can remove this condition after we ensure that this situation is not possible.

Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/DarcLib/Helpers/ManifestHelper.cs:169]: ok, I see. We can return to that after we have a published VMR build

**Pull request merged**